_drgn.pyi

```diff
@@ -202,13 +203,27 @@ class Program:
         :attr:`SymbolBinding.WEAK` symbol is found, it is returned. Otherwise,
         any matching symbol (e.g., :attr:`SymbolBinding.LOCAL`) is returned. If
         there are multiple matching symbols with the same binding, one is
-        returned arbitrarily.
+        returned arbitrarily. To retrieve all symbols with this name, regardless
+        of their binding scope, use :meth:`Program.symbols()`.
 
         :param address_or_name: Address or name.
         :raises LookupError: if no symbol contains the given address or matches
             the given name
         """
         ...
+    def symbols(self, name: Optional[str]) -> List[Symbol]:
```

Comment:
Yes, great idea. I hadn't realized about the overlapping symbols. I can do it as part of the same C loop and callback, since the address and size are available. That would work for the Python API but the C API becomes a bit clunky. You'd either need to specify via flags what you want to search, or you'd need to provide several functions. Do you have a preference between the two APIs sketched below? (or a better alternative)
```c
/* all in one. bonus: could be extended for other search params */
enum {
    DRGN_SEARCH_NAME=(1 << 0)
    DRGN_SEARCH_ADDRESS=(1 << 1)
};
struct drgn_symbol_search {
    const char *name;
    uint64_t address;
};
LIBDRGN_PUBLIC struct drgn_error *
drgn_program_search_symbols(struct drgn_program *prog,
			          const struct drgn_symbol_search *params,
			          unsigned int flags,
				  struct drgn_symbol ***syms_ret,
				  size_t *count_ret);

/* multiple C functions */
LIBDRGN_PUBLIC struct drgn_error *
drgn_program_find_symbols_by_name(struct drgn_program *prog,
			          const char *name,
				  struct drgn_symbol ***syms_ret,
				  size_t *count_ret);
LIBDRGN_PUBLIC struct drgn_error *
drgn_program_find_symbols_by_address(struct drgn_program *prog,
			          uint64_t address,
				  struct drgn_symbol ***syms_ret,
				  size_t *count_ret);
```

There's not much to gain by searching for name AND address, so either works. If you think it may be useful to lookup all global symbols, or all symbols per module, or all symbols of a type (data, function, etc), then the first would be able to suit any combination of those. (not sure whether the semantic would be AND or OR)

Reply:
I'd prefer the separate functions for now, for consistency with `drgn_program_find_symbol_by_address` and `drgn_program_find_symbol_by_name`.